### PR TITLE
Clean up `Note` classes

### PR DESCRIPTION
**Pull request by danielbitzer**
### Changes proposed in this Pull Request:

We need to add some more inbox notifications (#860) and while I was looking into it there seemed to be a number of duplicated lines in each of the `Automattic\WooCommerce\GoogleListingsAndAds\Notes\*` classes. Before adding any new notes classes I wanted to clean this up.

This PR also prevents direct access to these files.

**Question:** I don't see that these classes are using `OptionsAwareInterface` at all. Am I missing something?

### Testing

This is a little tricky since each note has different criteria. Deleting your existing notes `DELETE FROM wp_wc_admin_notes WHERE name LIKE 'gla-%'` would be the first step.

### Changelog entry

>

### CODE REVIEW

File: src/Notes/Note.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\Notes;
+
+use Automattic\WooCommerce\Admin\Notes\DataStore;
```
Comment (layoutd) [inline src/Notes/Note.php]: `DataStore` is unused.

**Review by layoutd — APPROVED** [review]
I tested all four existing notes and they were displayed correctly with their respective criteria. One small unused import to remove from the new abstract class, and a few other bits and bobs to address, but looks good to merge.

- Regarding the `OptionsInterface` question, I'm guessing it was either copied over from another class as a template, or was originally used or going to be used and then go refactored out (`HelperTraits::Utilities` has a few methods only used in the Notes classes that I'm assuming were extracted at some point). I'd say pull 'em out!

- A few comments regarding `CompleteSetup.php` (unrelated to your changes but since we're in the neighborhood)…
    - Line 67:
        > `* Check if it is > 3 days ago from DATE OF START OF SETUP (installation date)`

        This doesn't seem to be [a criteria anymore](https://github.com/woocommerce/google-listings-and-ads/commit/41c16313e8e0ee77fbcaff4b29c2d1be4033ff72#diff-7cebe8ef86c9f8e3c886840da6fb713ecc52c220c6967a71fa5a57b2254815b0), can probably remove the line from the documentation.

    - Line 81 - it's also a bit weird that it checks for EXACTLY five orders, I wonder if that was supposed to be five **or more** orders? @jconroy

**Comment by danielbitzer** [comment]
Thanks for the speedy review @layoutd! Also thanks for the clarification about the `OptionsAwareInterface`, I have removed them.

I'm going to spin your last comment into a new issue for discussion.

**Comment by layoutd** [comment]
**TLDR**: the individual notes classes _don't_ need to have `OptionsAwareTrait` inserted or implement `OptionsAwareInterface`, but the new abstract `Note` class _does_ need to implement `OptionsAwareInterface`. I made the change in 09c8ced694c039f6f71a5cb74b8c1ff4e57696c7

---

Mea culpa - looks like `OptionsAwareInterface` was actually necessary:
```
Fatal error: Uncaught Error: Call to a member function get() on null in /…/google-listings-and-ads/src/HelperTraits/Utilities.php on line 62
```

The `OptionsAwareTrait` is inherited by the `Utilities`, but the actual note classes (or, specifically, the abstract `Note` class) _do_ need to implement `OptionsAwareInterface` in order for the `options` property to be initialized correctly.

OptionsAwareTrait has the `set_options_object` method (required by the `OptionsAwareInterface`):
https://github.com/woocommerce/google-listings-and-ads/blob/2b7d4e4dd9dd860168aa194497f26d10d0d1f690/src/Options/OptionsAwareTrait.php#L27-L29
_Error thrown in `develop`_

which is invoked for all classes that implement the `OptionsAwareInterface` by inflection:
https://github.com/woocommerce/google-listings-and-ads/blob/3365b5ce38af88fd7b12ff9318b0a49069d25a36/src/Internal/DependencyManagement/CoreServiceProvider.php#L177-L179

**Comment by danielbitzer** [comment]
@layoutd good catch! 

> The `OptionsAwareTrait` is inherited by the `Utilities`, but the actual note classes (or, specifically, the abstract `Note` class) _do_ need to implement `OptionsAwareInterface` in order for the `options` property to be initialized correctly.

This seems like quite an opaque OOP structure to me. I feel like it may not be ideal to use a `{Dependency}AwareTrait` in a another trait due to the interface requirement. It's also a hard to find error because it doesn't appear upon compiling but rather it requires `$this->options` to be used before it throws an error.

Perhaps @JPry could shed some light here?

**Comment by JPry** [comment]
> > The `OptionsAwareTrait` is inherited by the `Utilities`, but the actual note classes (or, specifically, the abstract `Note` class) _do_ need to implement `OptionsAwareInterface` in order for the `options` property to be initialized correctly.
> 
> This seems like quite an opaque OOP structure to me. I feel like it may not be ideal to use a `{Dependency}AwareTrait` in a another trait due to the interface requirement. It's also a hard to find error because it doesn't appear upon compiling but rather it requires `$this->options` to be used before it throws an error.
> 
> Perhaps @JPry could shed some light here?

The original idea I had with `{Dependency}AwareTrait` and `{Dependency}AwareInterface` classes was they would provide simple helper methods at the Container level without explicitly requiring all of the classes to declare those dependencies in the constructor. This leaves the constructor with the dependencies that are more unique to what the class needs. In theory, it's also easy to see what "awareness" a class has directly.

I agree that things start to get a bit tricky when traits start including other traits. There might be times where this makes sense, but I think we should probably keep those instances to an absolute minimum. Ideally we should try to eliminate those altogether.

For the `Utilities` trait in particular, it might make more sense to remove the dependency of the `OptionsAwareTrait` entirely. It might make more sense to convert the `Utilities` trait to its own class, but I would be open to other ideas as well.

**Comment by danielbitzer** [comment]
Thanks for that info @JPry! I'm still learning my way around Woogle so that's very helpful.
